Update bounds on ghc-exactprint, reimplement overlap check

xich opened the pull request:
Bumped the bounds so apply-refact compiles with newest ghc-exactprint.

Also, the overlap check had a bug. Consider the following two ideas given by hlint:

```
(S1, [R1, D1])
(S2, [R2, D2])
```

(where Sn = idea message, Rn = a Replace constructor, Dn = a Delete constructor)

If R1 overlapped with R2 or D2, it would previously get filtered out, leaving [D1, R2, D2]... essentially only half of idea 1 would get implemented, and code would break. I reimplemented it to consider all the refactorings of a given idea as a unit.

mpickering commented:
Thank you @xich. These changes look good and the bug makes sense. I will get to review this on Saturday at the latest. 

The old overlap checker preferred the largest hint when there were overlapping hints. Do you think this is good behaviour? I could never decide for myself.

xich commented:
This is probably different than before... but I think, in general, picking the larger edit is probably better, especially if a hint eliminates code.

It now picks the *first* in a set of overlapping hints. If two hints both start in the same place, it picks the larger one. Also, if you have [R1,R2,R3] and R2 overlaps with both R1 and R3, but R1 is disjoint from R3, this will discard R2, then leave R3 (because it is no longer overlapping). That is a bit subtle, but it seems okay. Before it was possible that both R1 and R3 would get discarded if R2 was bigger than R1. Now it will always discard R2.

I actually started by modifying the original, because HLint was eta-reducing the following hint:

```
error = anyM (return . f) xs ==> return (any f xs)
```

and giving two ideas for it (one with `xs` eta-reduced). apply-refact was choosing the eta-reduced idea for some reason (which should have been smaller). This wouldn't really be a problem, except the refactorings coming out of HLint didn't always include new parens when necessary, so applying the eta-reduced rule would sometimes lead to compile errors, whereas the non-eta-reduced version would work fine. Then I noticed the bug I mentioned above, so ended up rewriting it anyway. :-P

mpickering commented:
The eta-reduce hint is quite badly broken anyway. It will drop local binds, I need to get Neil to disable it until I come up with a better implementation.

mpickering commented:
Some of the tests fail because of this PR. One of the failures is because ghc-exactprint changed (the one with the case statement). Could you please update the output for the other two please?

xich commented:
Yep, I can do that.

Design question... would you consider moving the `summary` function I wrote in the `refact` package itself? I find myself needing the same thing within `hlint` itself. (I'm fixing it to move comments if they would be deleted by a `Refactoring`.)

mpickering commented:
No - I don't think so. Can you give me a specific example of a case you are trying to fix? It seemed to me if you want to delete something then you should also remove the comments. 

The problem is that `ghc-exactprint` doesn't have a smart way of associating the comments to the right place. It associates then with the immediately following AST element, no matter what that is. I've tried various ways to reallocate the comments but nothing seems to work too well. It seems a sensible rule is, if the comment is on a newline, associate it with the *next* AST element, if not then associate it with the previous AST element. Even with this rule there are corner cases.. If you have a comment like

```
case foo x of
  True -> False
  False -> True -- My comment
```

then do we want to associate the comment with either
1. The expression `True`
2. The branch `False -> True`
3. The whole case expression.

They all seem to be sensible answers but in fact the answer is 4. Whatever comes next. 

If someone could come up with something more sensible then that would be much better. It seems like I lost the code which attempted to do this but I loosely based it on the code in hindent.

xich commented:
Ah, that will work! I just expected comments to be full-fledged AST nodes themselves. I didn't understand that they were attached to another expression.

For context, I implemented the following hint in hlint and am using it to rewrite:

```haskell
error = do x <- y; return (f x) ==> f <$> y
```

The problem is when a comment is somewhere in there, like this:

```haskell
do x <- foo
    {- some comment -}
    return (bar x)
```

HLint has access to a raw list of comments, but they aren't included in its AST type. I'm trying to just spot comments which overlap with a refactoring and move them, like his:

```haskell
do {- some comment -}
    bar <$> foo
```

I've modified hlint to generate `InsertComment` refactorings when they overlap with replace/delete refactorings... I just needed to figure out how to generate the SrcSpan.

Thanks for the help!

mpickering commented:
I don't think that adding "InsertComment" is the right solution but would
need to see the code.

On Fri, Oct 9, 2015 at 8:18 AM, Andrew Farmer <notifications@github.com>
wrote:

> Ah, that will work! I just expected comments to be full-fledged AST nodes
> themselves. I didn't understand that they were attached to another
> expression.
>
> For context, I implemented the following hint in hlint and am using it to
> rewrite:
>
> error = do x <- y; return (f x) ==> f <$> y
>
> The problem is when a comment is somewhere in there, like this:
>
> do x <- foo
>     {- some comment -}
>     return (bar x)
>
> HLint has access to a raw list of comments, but they aren't included in
> its AST type. I'm trying to just spot comments which overlap with a
> refactoring and move them, like his:
>
> do {- some comment -}
>     bar <$> foo
>
> I've modified hlint to generate InsertComment refactorings when they
> overlap with replace/delete refactorings... I just needed to figure out how
> to generate the SrcSpan.
>
> Thanks for the help!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mpickering/apply-refact/pull/8#issuecomment-146775624>
> .
>

xich commented:
I updated the test output for the two tests that were broken by behavior change of the overlap check. In both cases, there were two refactorings which started in the same position. The old check chose the first one, the new check chooses the longest one. In both cases, the longest one was not the first one, so a different refactoring got applied.

I didn't update the output for the test broken by the new `ghc-exactprint` because I wasn't sure if that was expected, or pointed to a bug somewhere. It seems strange to lose the indentation on the case alternatives.

mpickering commented:
Yes thanks, it is a known bug in `ghc-exactprint` which I haven't had time to look at yet. The Haskell eXchange, GHC patches and the start of the school year have made it a very busy time for me.

alanz commented:
wrt the bug in ghc-exactprint

The indentation of the alternatives for a case was broken before, and that breakage has been undone. Unfortunately this has exposed a different bug which needs to be addressed.

As I understand things.

xich commented:
Do either of you have a general idea of where things might be going wrong? I could maybe spend some cycles on it.

alanz commented:
If I recall it is related to the alts in a case being captured relative to the start of the case statement, vs relative to the start of the prior line.

So

```haskell
foo x = case (baz x) of
    1 -> "a"
    _ -> "b"
```

becomes

```haskell
fooLonger x = case (baz x) of
         1 -> "a"
         _ -> "b"
```

instead of

```haskell
fooLonger x = case (baz x) of
    1 -> "a"
    _ -> "b"
```

The bug fixed sorted out

```haskell
foo x = case (bazLonger x) of
    1 -> "a"
    _ -> "b"
```

not moving the position of the alts

alanz commented:
Well, I added my example test cases to ghc-exactprint via https://github.com/alanz/ghc-exactprint/commit/82c99d2ff6a5851c265b86a366bae365ce2c5121 but they both pass.

@mpickering can you clarify what the ghc-exactprint bug is?

xich commented:
@alanz The test/examples/RedundantDo.hs test is failing.

Original: https://github.com/mpickering/apply-refact/blob/master/tests/examples/RedundantDo.hs
Expected: https://github.com/mpickering/apply-refact/blob/master/tests/examples/RedundantDo.hs.expected

What is now happening: https://github.com/mpickering/apply-refact/blob/30424a7704de1543f1f84966b16eb9e7f87e4f7a/tests/examples/RedundantDo.hs.expected

Which actually was the original expected output of the test, but it was updated to the current expectation (with indented alternatives) here c6c4aa081231d11d66cb1dfff122c47018401e63. So it looks like ghc-exactprint 0.3.1 fixed it, then 0.4.1 broke it again.

alanz commented:
I am not all that familiar with the innards of apply-refact.

I do know that there used to be an error in ghc-exactprint which meant that
the top left location of an AST fragment did not serve as an anchor for the
rest to flow when there was modification. This was corrected as seen by the
two new tests I just added to ghc-exactprint.

There is something more complicated going on with the redundant do example,
which is an interaction between the layout captured for the do which sets a
relative starting column for the case, and the version when it is removed.
In  that case I am not sure what is used as the starting column for working
out the indentation. Sorry if this all sounds a bit hand-wavy.

I will not be able to help/look further right now, and will be travelling
until Sunday.

What I do suggest is to see what the ParsedSource looks like before
modification, after modification, and what it should look like if you just
feed it the expected result.  The best way `of doing that is to use
`showAnnData` from [1] or the one wrapped in the Transform monad
`logDataWithAnnsTr` [2]. Or just run `ghc-hare -v roundtrip FILENAME" and
it will dump the AST in a file with the suffix "AST_out"

[1]
https://github.com/alanz/ghc-exactprint/blob/master/src/Language/Haskell/GHC/ExactPrint/Utils.hs#L325
[2]
https://github.com/alanz/ghc-exactprint/blob/master/src/Language/Haskell/GHC/ExactPrint/Transform.hs#L155

On Wed, Oct 14, 2015 at 11:15 PM, Andrew Farmer <notifications@github.com>
wrote:

> @alanz <https://github.com/alanz> The test/examples/RedundantDo.hs test
> is failing.
>
> Original:
> https://github.com/mpickering/apply-refact/blob/master/tests/examples/RedundantDo.hs
> Expected:
> https://github.com/mpickering/apply-refact/blob/master/tests/examples/RedundantDo.hs.expected
>
> What is now happening:
> https://github.com/mpickering/apply-refact/blob/30424a7704de1543f1f84966b16eb9e7f87e4f7a/tests/examples/RedundantDo.hs.expected
>
> Which actually was the original expected output of the test, but it was
> updated to the current expectation (with indented alternatives) here
> c6c4aa0
> <https://github.com/mpickering/apply-refact/commit/c6c4aa081231d11d66cb1dfff122c47018401e63>.
> So it looks like ghc-exactprint 0.3.1 fixed it, then 0.4.1 broke it again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mpickering/apply-refact/pull/8#issuecomment-148202981>
> .
>

mpickering commented:
Here is the commit which made the test break - https://github.com/alanz/ghc-exactprint/commit/622ae957432351f9a480fb9c67e979fb4328271f

Useful comment explaining: https://github.com/alanz/ghc-exactprint/commit/1947241a0418481991f9a55561186e38245cfab6


So as I see it: It was correct after my changes, a different problem caused Alan's test case to fail and the revert broke my test.

Here is the AST from his test case.

```

({ test.hs:1:1 }
 Just (Ann (DP (0,0)) [] [] [((G AnnEofPos),DP (1,0))] Nothing Nothing)
 (HsModule 
  (Nothing) 
  (Nothing) 
  [] 
  [
   ({ test.hs:(1,1)-(3,10) }
    Just (Ann (DP (0,0)) [] [] [] Nothing Nothing)
    (ValD 
     (FunBind 
      ({ test.hs:1:1-3 }
       Just (Ann (DP (0,0)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
       (Unqual {OccName: foo})) 
      (False) 
      (MG 
       [
        ({ test.hs:(1,1)-(3,10) }
         Just (Ann (DP (0,0)) [] [] [((G AnnEqual),DP (0,1))] Nothing Nothing)
         (Match 
          (Just 
           ((,) 
            ({ test.hs:1:1-3 }
             Just (Ann (DP (0,0)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
             (Unqual {OccName: foo})) 
            (False))) 
          [
           ({ test.hs:1:5 }
            Just (Ann (DP (0,1)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
            (VarPat 
             (Unqual {OccName: x})))] 
          (Nothing) 
          (GRHSs 
           [
            ({ test.hs:(1,7)-(3,10) }
             Just (Ann (DP (0,-1)) [] [] [] Nothing Nothing)
             (GRHS 
              [] 
              ({ test.hs:(1,9)-(3,10) }
               Just (Ann (DP (0,1)) [] [] [((G AnnCase),DP (0,0)),((G AnnOf),DP (0,1))] Nothing Nothing)
               (HsCase 
                ({ test.hs:1:14-20 }
                 Just (Ann (DP (0,1)) [] [] [((G AnnOpenP),DP (0,0)),((G AnnCloseP),DP (0,0))] Nothing Nothing)
                 (HsPar 
                  ({ test.hs:1:15-19 }
                   Just (Ann (DP (0,0)) [] [] [] Nothing Nothing)
                   (HsApp 
                    ({ test.hs:1:15-17 }
                     Just (Ann (DP (0,0)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
                     (HsVar 
                      (Unqual {OccName: baz}))) 
                    ({ test.hs:1:19 }
                     Just (Ann (DP (0,1)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
                     (HsVar 
                      (Unqual {OccName: x}))))))) 
                (MG 
                 [
                  ({ test.hs:2:3-10 }
                   Just (Ann (DP (1,-6)) [] [] [] Nothing Nothing)
                   (Match 
                    (Nothing) 
                    [
                     ({ test.hs:2:3 }
                      Just (Ann (DP (0,0)) [] [] [] Nothing Nothing)
                      (NPat 
                       ({ test.hs:2:3 }
                        Just (Ann (DP (0,0)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing){HsOverLit:1}) 
                       (Nothing) 
                       (HsLit 
                        (HsString "" {FastString: "noSyntaxExpr"}))))] 
                    (Nothing) 
                    (GRHSs 
                     [
                      ({ test.hs:2:5-10 }
                       Just (Ann (DP (0,1)) [] [] [((G AnnRarrow),DP (0,0))] Nothing Nothing)
                       (GRHS 
                        [] 
                        ({ test.hs:2:8-10 }
                         Just (Ann (DP (0,1)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
                         (HsLit 
                          (HsString "\"a\"" {FastString: "a"})))))] 
                     (EmptyLocalBinds)))),
                  ({ test.hs:3:3-10 }
                   Just (Ann (DP (1,0)) [] [] [] Nothing Nothing)
                   (Match 
                    (Nothing) 
                    [
                     ({ test.hs:3:3 }
                      Just (Ann (DP (0,0)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
                      (WildPat 
                       (PlaceHolder)))] 
                    (Nothing) 
                    (GRHSs 
                     [
                      ({ test.hs:3:5-10 }
                       Just (Ann (DP (0,1)) [] [] [((G AnnRarrow),DP (0,0))] Nothing Nothing)
                       (GRHS 
                        [] 
                        ({ test.hs:3:8-10 }
                         Just (Ann (DP (0,1)) [] [] [((G AnnVal),DP (0,0))] Nothing Nothing)
                         (HsLit 
                          (HsString "\"b\"" {FastString: "b"})))))] 
                     (EmptyLocalBinds))))] 
                 [] 
                 (PlaceHolder) 
                 (FromSource))))))] 
           (EmptyLocalBinds))))] 
       [] 
       (PlaceHolder) 
       (FromSource)) 
      (WpHole) 
      (PlaceHolder) 
      [])))] 
  (Nothing) 
  (Nothing)))
```

The problem is this bit..

```
({ test.hs:2:3-10 }
                   Just (Ann (DP (1,-6)) [] [] [] Nothing Nothing)
                   (Match 
                    (Nothing) 
                    [
```

The delta is `(1, -6)` when it should be `(1, 2)`. I think this should be an easy fix in `ghc-exactprint` but haven't got around to looking myself yet. That along with reverting the first commit I pointed to should get both test cases to work. I can give a more complete description to why I think that is the problem tomorrow but that should be enough to get started if you want to hack around before then.

mpickering commented:
I think what I said is actually wrong. I added a failing test to my branch of ghc-exactprint anyway.

mpickering commented:
@xich maybe you can add a different perspective here.

Do you think the behaviour which alan described [earlier](https://github.com/mpickering/apply-refact/pull/8#issuecomment-148152699) in the ticket is desirable? I can see how it might be but it violates a key principle what you should be able to move the top left of an expression and the rest of the expression should print out exactly the same way as before. 

To be concrete, I am proposing if the case keyword moves then so should the branches so whatever happens, after the transformation they are still in the same place relative to each other.

mpickering commented:
I am going to merge this now. Thanks @xich!
